doc/cookbook/custom_dna4.cpp

```diff
@@ -0,0 +1,58 @@
+#include <seqan3/alphabet/nucleotide/nucleotide_base.hpp>
+#include <seqan3/core/debug_stream.hpp>
+
+class my_dna4 : public seqan3::nucleotide_base<my_dna4, 4>
+{
+public:
+    using nucleotide_base<my_dna4, 4>::nucleotide_base; // Pull constructors of the base class into public scope.
+
+private:
+    // Value to char conversion table.
+    static constexpr char_type rank_to_char[alphabet_size] {'A', 'C', 'G', 'T'}; // rank 0,1,2,3
+
+    // Char-to-value conversion table. This is where char conversion for assign_char() is handled!
+    static constexpr std::array<rank_type, 256> char_to_rank
+    {
+        [] () constexpr
+        {
+            std::array<rank_type, 256> conversion_array{}; // By default now, everything has rank 0 which equals `A`.
+
+            conversion_array['C'] = conversion_array['c'] = 1;
+            conversion_array['G'] = conversion_array['g'] = 2;
+            conversion_array['T'] = conversion_array['t'] = 3;
+            conversion_array['U'] = conversion_array['T']; // set U equal to T
+            conversion_array['u'] = conversion_array['t']; // set u equal to t
+
+            return conversion_array;
+        }()
+    };
+
+    // The forward declaration of the complement table. With this, my_dna4 models seqan3::nucleotide_alphabet.
+    static const std::array<my_dna4, alphabet_size> complement_table;
```

Comment:
Why is the conversion table defined out of class but the char to rank not? And above you call it conversion array but here table. I’d like to have them named equally for consistency reasons.

Reply:
I copied from dna4 so I am not 100% sure but I think it is because the complement table stores instances of the class itself so declaring it within the class results in an infinite loop of trying to define the class?

Reply:
My take was, that you can't instantiate a static class member within the class itself.

I might be wrong about this...

Reply:
Yes, but an inline static member can be defined in class.

Reply:
Ah ok, now I get it. I overlooked the `my_dna4` type inside of the array declaration. Of course, we cannot define it inside of the class because it is not a fully defined type yet. I am wondering if foward declaring it would solve this? Anyway, a comment should be added, because users might ask the same question. It is obviously not immediately clear why both arrays are defined differently.